[2.0.x]  Add L6470 SPI daisy chain support

Bob-the-Kuhn opened the pull request:
This PR adds SPI daisy chain support for the L6470 stepper drivers.  The following file contains an explanation of the hardware and software involved.
[two_motors_synchronized.txt](https://github.com/MarlinFirmware/Marlin/files/2752825/two_motors_synchronized.txt)

Arduino-L6470 library revision 0.7.0 is required for this stepper driver.  This library has not been released yet.

The L6470 drivers continue to operate in STEP_CLOCK mode.  In this mode the direction and enable are done via SPI commands and the phase currents are changed in response to step pulses (generated in the usual way).

**HARDWARE/SOFTWARE interaction**
Powering up a stepper and setting the direction are done by the same command.  Can't do one without the other.

**All** directions are set **every time** a new block is popped off the queue by the stepper ISR.

SPI transfers when setting the directions is minimized by using arrays and a SPI routine dedicated to this function.  L6470 library calls are not used.  For N L6470 drivers, this results in a N byte transfer.  If library calls were used then N*N bytes would be sent.

Power up (reset) sequence:
- Stepper objects are created before the **setup()** entry point is reached.
- The L6470_chain array is populated during **setup()**. This array is used to tell where in the SPI stream the commands/data for an stepper is positioned. 
- The L6470 soft SPI pins are initialized.  
- The L6470 chips are initialized during **setup()**.   They can be re-initialized using the **L6470_init_to_defaults()** function
- The steppers are **NOT** powered up during this sequence.

**Top level changes:**
- Change all occurrences of L6470 to ST_L6470 to resolve name conflict with the library.
- Attempted to emulate some of the TMC functionality by adding a common code module  (L6470_Marlin), a monitor function and custom gcodes.
- The majority of the changes were in stepper_indirection and in stepper.  These were to modify/update the setup of the stepper objects and support enabling/disabling the drivers.  Code was added to minimize the time (number of SPI transmissions) when setting direction.
- A fastio based soft SPI (HAL_spi_L6470.cpp) was added to support the SPI daisy chain transfers.  Since it is fastio based it can be used by all HALs so it is in the HAL/shared folder.

The numbering of the gcodes will probably change.  I didn’t find anything comparable (except for M906) so the assignment of numbers was arbitrary.

The example configurations have only had the L6470 to ST_L6470 name changes done.  I’ll update them once this PR has had some feedback.

---

**Marlin/Configuration_adv.h**
- Added daisy chain specific  SPI pin assignments and chain positions
- The 0.7.0 library supports SPI pins assignments for individual drivers.  Pin assignments were added for these.  
- Added a “MAX_VOLTAGE” define for each driver.  This is used to initialize a register called KVAL_HOLD.  This register sets the max duty cycle for the stepper voltage PWMs which sets the power used by the stepper.  This setting is critical in tuning the system to get maximum performance without over heating the drivers.

**HAL_spi_L6470.cpp** contains two SPIs and a pin init routine.  

- One routine inserts the data of interest in the correct position in the SPI transfer sequence and sends NOOPs to all the other drivers.  This one is used to transfer data to/from an individual driver such as during setup.  In a 6 driver system each transfer consists of one byte for each driver.  It can take quite some time to change settings using this routine. 
- The other routine is meant to minimize the time needed to set the direction and enable the drivers when a new planner block is popped off the queue.  It just transmits a buffer .  The buffer contents are filled by Marlin (the library is not involved).  

**Marlin.cpp**
- Added running, if enabled, a L6470 monitor to the manage_inactivity() routine.
- Modified manage_inactivity() routine to only run the stepper shutdown routine once when the shutdown timer expires.  Previously it ran the shutdown routine every time it ran. When a L6470 driver was present it looked like 99% of the free time was spend doing SPI transfers with the old method.
- Moved the stepper driver chip initialization/setup into the SETUP portion.  In the old location the SPI transfers were being attempted before the peripherals were initialized so nothing happened.

**drivers.h**
Added the macro AXIS_IS_ST_L6470(A)

**enum.h**
Added L6470_driver_enum to provide unique names/indices for each of the 13 possible drivers.

**macros.h**
Defined macro UNUSED(x) to eliminate a compiler warning.

**M114**
Displays the absolute positions registers of the L6470 drivers during a M114 D.  They should be within 1 of the counts for each axis.

**G28**
Set the absolute positions registers of the L6470 drivers to the counts for each axis after homing completes.

**Added gcodes**
- M122 - Added L6470 version of M122 to report status of each driver.
- M906 – set/display KVAL_HOLD register.  Allows user to do run time tuning.
- M916: L6470 tuning: Increase drive level until get thermal warning
- M917: L6470 tuning: Find minimum current thresholds
- M918: L6470 tuning: Increase speed until max or error

**Conditionals_post.h**
Macro **USE_EXECUTE_COMMANDS_IMMEDIATE** modified to be true when there is a L6470 driver in the system.  The M916-M918 routines use this feature to inject movement commands into the queue.

**SanityCheck.h**
**[AXIS]_IS_L6470** is now check modified to use ST_L6470

**L6470_Marlin.h**
Contains all the non-library definitions needed to support the L6470 software.

**L6470_Marlin.cpp**
Contains the common code and data structures.

### stepper.cpp
routine **set_directions()** modified to generate and xmit a buffer of L6470 commands to set each driver to the correct direction.  This also enables the driver.  In a N driver system this results in sending N bytes when setting/changing direction rather the N squared bytes.

**stepper_indirection.cpp**
- Does a hardware reset of all drivers if the reset pin exists.  This is done just once before writing to any of the registers in any of the chips.
- Added chain setup, individual driver SPI pin setup, KVAL_HOLD setting and clears out the absolute position register when initializing individual drivers..
- Added axis level disable routines.

**stepper_indirection.h**
UIpdated driver enable macros.


**Misc changes**
added date stamp to AutoBuild script


**TO DO:**
~Add  MIXING_EXTRUDER section for L6470 drivers to **stepper_indirection.h**~ **DONE** - created DIR_WRITE macros such that MIXING_EXTRUDER section did not have to be modified.
~MIXING_EXTRUDER doesn't matter to a L6470 system because all L6470 drivers always have direction set and enabled simultaneously~    **WRONG** - In some extruder systems the directions are not all the same for every extruder.


~Update example configurations~ **DONE**
~Update pinsDebug_list.h~  **DONE**
Test in mixed driver system

Bob-the-Kuhn committed:
Add L6470 SPI daisy chain support

Bob-the-Kuhn commented:
**Some big assumptions:**
- ~All **X** drivers are the same (either all L6470 or all something else)~
- ~All **Y** drivers are the same (either all L6470 or all something else)~
- ~All **Z** drivers are the same (either all L6470 or all something else)~
- ~All **extruder** drivers are the same (either all L6470 or all something else)!~

thinkyhead committed:
Move strings to PROGMEM, and other fixes

thinkyhead committed:
Restore ST_L6470 to L6470

- Avoid L6470 library name conflict by another method

thinkyhead commented:
I'm a maintainer of the Arduino-L6470 library. Should I do a formal release of 0.7.0?

thinkyhead commented:
> Change all occurrences of L6470 to ST_L6470 to resolve name conflict with the library.

I modified the macros in `driver.h` so that the symbol `L6470` is no longer declared, preventing the naming conflict with the `L6470` class.

thinkyhead commented:
Note that this PR breaks mixing, dual-x duplication mode, and other special extrusion modes for the L6470. To fix this the L6470 macros to enable/disable stepper drivers should be extended for all the extrusion modes.

Bob-the-Kuhn commented:
I think I've gotten the L6470 extruder macros setup so that the standard mixing extruder logic works correctly in a L6470 system.

Bench testing shows that the new defines are working.  Lots more testing to see if that's true in all cases.

I'm going to try to squash the commits.

Bob-the-Kuhn commented:
> I'm a maintainer of the Arduino-L6470 library. Should I do a formal release of 0.7.0?

Not yet.

I need to do some more bench testing and then actually try a print before stating that it's ready for public consumption.

Bob-the-Kuhn commented:
I can't reproduce the Travis errors.  Is Travis having problems?

thinkyhead commented:
I've squashed and pushed the commits, so you can just do this:
```
git fetch origin
git checkout L6470-PR
git reset --hard origin/L6470-PR
```

thinkyhead commented:
> Is Travis having problems?

The issue is weird. I'm trying a small change to see if it finally compiles. There was something else blocking it before this most recent "TMCTMC2130Stepper WTF" compile error.

Bob-the-Kuhn committed:
Fixes following cleanup

Bob-the-Kuhn commented:
Thanks for all your work.

To get it to compile for a L6470 driver I needed to fix a few syntax errors and had to revert the argument definitions in **L6470_get_user_input**.

I'm trying to keep the steppers powered down after a reset.  Currently **configuration_store.cpp** calls **reset_stepper_drivers()** which calls **set_directions()** which powers up the L6470s .  About 1/2 second later they get disabled.

To avoid that 1/2 second power up I've modified **reset_stepper_drivers()** to call **set_directions()** only if a TMC driver is in the system.  

Why does **configuration_store.cpp** call **reset_stepper_drivers()** ?  Maybe there's a better way to accomplish the goal.

Bob-the-Kuhn commented:
Some more changes.

**HAL_spi_L6470.cpp**.
- Removed test code from **HAL_spi_L6470.cpp**.
- Reverted arguement list for **L6470_Transfer()**.  The library doesn't know about **pin_t** so the definitions didn't match which meant the dummy function was used instead of the desired one.

**M122.cpp** 
- Fixed a logic error.

**L6470_Marlin.cpp**
- Reverted **L6470_monitor_update** because of compile errors with **append_stepper_err**
```
In file included from C:\Users\bobku\.platformio\packages\framework-arduinosam\cores\sam/Arduino.h:31:0,

from Marlin\src\module\L6470\../../inc/../HAL/HAL_DUE/HAL.h:34,

from Marlin\src\module\L6470\../../inc/MarlinConfig.h:30,

from Marlin\src\module\L6470\L6470_Marlin.cpp:27:

Marlin\src\module\L6470\L6470_Marlin.cpp: In function 'void append_stepper_err(char*&, uint8_t, const char*)':

C:\Users\bobku\.platformio\packages\framework-arduinosam\cores\sam/avr/pgmspace.h:99:59: error: expected primary-expression before ')' token

#define sprintf_P(s, f, ...) sprintf((s), (f), __VA_ARGS__)

^

Marlin\src\module\L6470\L6470_Marlin.cpp:586:19: note: in expansion of macro 'sprintf_P'

if (err) p += sprintf_P(p, err);

^

Marlin\src\module\L6470\L6470_Marlin.cpp: In function 'void L6470_monitor_update(uint8_t, uint16_t)':

Marlin\src\module\L6470\L6470_Marlin.cpp:603:11: error: invalid operands of types 'char*' and 'void' to binary 'operator+'

p += append_stepper_err(p, stepper_index, PSTR(" - communications lost\n"));

^

Marlin\src\module\L6470\L6470_Marlin.cpp:603:11: error:   in evaluation of 'operator+=(char*, void)'

Marlin\src\module\L6470\L6470_Marlin.cpp:610:13: error: invalid operands of types 'char*' and 'void' to binary 'operator+'

p += append_stepper_err(p, stepper_index, PSTR(" - still no communications\n"));

^

Marlin\src\module\L6470\L6470_Marlin.cpp:610:13: error:   in evaluation of 'operator+=(char*, void)'


```

Bob-the-Kuhn committed:
add new L6470 pins to pinsDebug_list.h

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -577,11 +577,11 @@
  * A4988 is assumed for unspecified drivers.
  *
- * Options: A4988, A5984, DRV8825, LV8729, L6470, TB6560, TB6600, TMC2100,
+ * Options: A4988, A5984, DRV8825, LV8729, ST_L6470, TB6560, TB6600, TMC2100,
```
This name change really isn't needed. There shouldn't be any naming conflict with the Arduino-L6470 library with the proposed change to `drivers.h`. I assume you reverted the name change so that you could bring in your additional changes, and not because it was a problem…?

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -2022,7 +2011,7 @@
  *  - Turn off after the print has finished and the user has pushed a button
  */
-#if ENABLED(BLINKM) || ENABLED(RGB_LED) || ENABLED(RGBW_LED) || ENABLED(PCA9632) || ENABLED(PCA9533)|| ENABLED(NEOPIXEL_LED)
+#if ENABLED(BLINKM) || ENABLED(RGB_LED) || ENABLED(RGBW_LED) || ENABLED(PCA9632) || ENABLED(NEOPIXEL_LED)
```
I'd like to keep all the above changes starting with "`// 2-wire Non-latching LCD`…". (A "git rebase" to the latest bugfix-2.0.x would've avoided needing to fix these throughout the PR. Possibly a rebase now will still work, so I'll try that.)

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -135,7 +135,7 @@
 
 // Show extra position information in M114
-//#define M114_DETAIL
+#define M114_DETAIL
```
We definitely don't want to include changes to the base `Configuration_adv.h` in a PR.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -1504,68 +1496,122 @@
+ * ST_L6470 Stepper Driver options
+ *
+ * The Arduino-6470 library (revision 0.7.0 and higher)  is required for this stepper driver.
+ * https://github.com/ameyer/Arduino-6470
```
The library is "`Arduino-L6470`" — without the "L" the name is incorrect and the link won't work.

thinkyhead commented:
Ok, so when you reverted my changes you also reverted a large number of unrelated files back to a point in time several days ago. I'm inclined to revert this back to the point in time before you reverted my changes, and ask you to make your additional changes starting from that point. Otherwise, trying to restore all the things that got reverted will take me many hours.

thinkyhead commented:
Ok, so I backed up to that point in time (f80dfc5) where all the configs were up to date and so on, and I was then able to cherry-pick your followup changes. You should now be able to do…
```
git fetch origin
git checkout L6470-PR
git reset --hard origin/L6470-PR
```
…for the rebased and squashed commits. L6470 name conflict should not be a problem, but hail me if a better patch is needed.

thinkyhead committed:
Additional cleanup

thinkyhead committed:
Use AXIS_DRIVER_TYPE_ macros for known axes

Bob-the-Kuhn commented:
Apparently I didn't do the git fetch origin … sequence properly.

Definitely want to minimize your work.  

After the
```
   git fetch origin
   git checkout L6470-PR
   git reset origin/L6470-PR
```

I ended up with a large commit (200+ files) so I just did a commit.  Maybe I should have discarded that commit. 

I'll check it out tomorrow.  Today I'm busily destroying a Geeetech ramps_FD-V1 trying to get my printer back up and running.  That board is one of the worst pieces of workmanship I've seen from China.  Broken/melted connectors, LEDs in backwards, wrong parts installed.

thinkyhead committed:
Reduce M906 code size

thinkyhead commented:
Oops! I should have written: `git reset --hard origin/L6470-PR`…
```
git fetch origin
git checkout L6470-PR
git reset --hard origin/L6470-PR
```

thinkyhead commented on Marlin/src/gcode/calibrate/G28.cpp:
```diff
@@ -440,4 +444,12 @@ void GcodeSuite::G28(const bool always_home_all) {
+
+  #if HAS_DRIVER(L6470)
+    // Set L6470 absolute position registers to counts
+    for (uint8_t j = 1; j <= L6470_chain[0]; j++) {
```
So, I see a lot of references to `L6470_chain` but I don't see a definition for it anyplace.

thinkyhead committed:
Do a test of L6470 driver type

thinkyhead commented:
I think this is ready, so whenever you have your `Arduino-L6470` PR for 0.7.0 ready to go let me know and I'll review it as quickly as possible.

gloomyandy commented:
Does this PR mean that SPI will now be used from an interrupt context? If so I think this is the first time that has happened? This will almost certainly have issues for any sort of shared spi usage (which may not be a problem here, but should probably be noted as in general with Marlin you can share spi pins).

Bob-the-Kuhn commented:
@gloomyandy - no, this is just another soft SPI.

Bob-the-Kuhn commented:
This last commit:
- removed unused variables
- corrected the status display logic in the L6470_monitor
- restored the DIR_WRITE macros to what we wanted
- corrected a couple of logic errors
- added hardware and software resets of the chips
- moved chitchat logic around

gloomyandy commented:
But doesn't the call made in set_directions() generate spi traffic during an interrupt call? It looks like set_directions is called from the stepper interrupt. If it does, that means you can't share the spi pins with other spi devices (that operate at non interrupt time). Also are you sure that there is no chance that the stepper interrupt could happen during some other non-interrupt time transfer to the L6470 (like monitoring the status)? Maybe I've misunderstood the code?

Bob-the-Kuhn commented:
Correct - when the stepper ISR is running the only time it generates L6470 SPI traffic is when it pops a new block off the queue and sets the directions.

Correct - this is a dedicated SPI.  No attempt has been made to allow sharing a SPI.  That could be done in the future.

> stepper interrupt could happen during some other non-interrupt time transfer to the L6470

Yes, this could happen.  I'll have to think about what could happen and how to guard against it.

Bob-the-Kuhn commented:
> stepper interrupt could happen during some other non-interrupt time transfer to the L6470

Just pushed some code that eliminates the possibility of the stepper ISR causing a garbage SPI command to be sent.
- Disabled interrupts during the transmission of a "non-set_directions() byte"
- Added handshake flags so that set_directions() can tell if it has interrupted a SPI transfer sequence and can then tell the interrupted code to stop transmitting
- Modified set_directions() to check the handshake flag and, if it's set,  a) set the abort other SPI sends flag and b) send enough NOOPs to complete any command that may have been interrupted.
- Modified SPI routine used by set_directions() to guarantee that the chip select timimg is correct.
- Modified all L6470 routines that could be interrupted by  set_directions() so that they set the handshake flag at the beginning, stopped transmitting if the abort flag is active and, at the end, cleared both the handshake flag and the abort flag.

Now to try figure out how to test this code ...

Bob-the-Kuhn commented:
I've re-written the enable/disable macros & I think that we can now mix different drivers within an axis.

gloomyandy commented:
> Now to try figure out how to test this code ...

Good luck! Sounds complex, does doing the set direction code via SPI have much impact on the stepper timing? It sounds like it will take a lot longer than just setting a pin!

thinkyhead commented:
@Bob-the-Kuhn — Please follow my instructions, use `git fetch` and `git reset --hard` as described before you apply additional changes to your old code and then force-push it, overwriting all my work and bringing back all the errors I'm trying to fix here.

thinkyhead committed:
Update L6470 configuration options

thinkyhead committed:
Bring in lost changes and Bob's new changes

thinkyhead committed:
Move L6470 code to 'libs' folder

thinkyhead commented:
I tried to patch together the changes you intended to make from the ones that seemed to be reversions. So, compare what you have in your working copy with this and see if everything looks ok.

Bob-the-Kuhn commented:
I think this PR is ready for public consumption.

My printer is still down but I was able to connect into some of its steppers.  With A4988 drivers on Z and E0 and L6470 drivers on Z2 and E1 everything worked as expected.  Both Z steppers turned the same way and the same amount.  The two extruders were independent and turned the same direction and same amount when selected.

As to the delay, using a DUE and a 6 stepper system, time from popping a new block from the queue to finish setting the directions is:
-  3.25uS when using only A4988 drivers
- 30.92uS when using only L6470 drivers

Bob-the-Kuhn commented:
I'm also playing with a powerSTEP01 driver.  It's a higher power version of the L6470 but there are enough differences that it'll be several more days before I have it fully functional..  There's changes in register bit definitions and register addresses and possibly setup/init settings.

Bob-the-Kuhn commented:
The last commits did the following:
- removed items specific to my printer from configuration_adv.h
- remove MAKEFILE from this PR (brought it up to the latest)

Bob-the-Kuhn commented:
@thinkyhead - sorry, I just saw your request to do the git fetch origin

Can you revert my changes?

Bob-the-Kuhn commented:
I've just copied the files from commit 91726c6d76220571e96841f5484c310229abfddb into this PR.  That should fix my hasty `git push -f`

Now to test.

Bob-the-Kuhn commented:
Scott's changes are now back in the PR and have been tested.

Had problems with the optional argument in L6470_say_axis() so made it required and updated the calls accordingly.

thinkyhead commented:
I'll need to pull this down, do a `git rebase` as usual, then do a `git diff` and try to cherry-pick the changes and make sure everything is matching up to expectations. I'll update the reorganized PR as soon as I get it together. It took me about 2 hours yesterday. I expect it will only take about an hour today.

thinkyhead committed:
Intentional config changes

thinkyhead committed:
Fix L6470_say_axis default argument

thinkyhead commented:
Ok, so this is almost ready, but before it's merged the `Arduino-L6470` 0.7.0 library will need to be published. That library update might still need a little work, but we can deal with any L6470 issues over on that repo.

thinkyhead committed:
Convert L6470 code to a singleton class

Bob-the-Kuhn commented:
> I'll need to pull this down, do a git rebase as usual, then do a git diff and try to cherry-pick the changes 

What??!!! How could you possibly think my amateur coding could be improved upon?   😉

thinkyhead commented:
Not much to improve, honestly! This is all pretty solid. The chain method is bound to be slow, so there's not much we can do about that.

The L6470 library has some interesting qualities, like requiring the client app (Marlin) to define `L6470_SPI_init` and `L6470_Transfer` on its behalf, otherwise weak-linking them as do-nothing functions. It probably does end up being smaller code than if (instead) we had to provide callback function pointers. But it's certainly an uncommon way to construct a library.

The way we would usually do it is to define the methods as pure-virtual, then require the client to make a sub-class `L6470` that defines these methods. Unfortunately that causes the code size to increase and slow down, since a virtual table needs to be added.

Weird as it is, I guess it'll be ok for our purposes.

thinkyhead committed:
L6470_chain => L6470::chain

Bob-the-Kuhn commented:
Thanks for all your work.  Much appreciated.

As to the "weak link" method - I was expecting someone to come back and point me to a better method.  Should I pursue making then a part of a class?
